flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/batch/connectors/hive/HiveTableSink.java

```diff
@@ -0,0 +1,159 @@
+ */
+public class HiveTableSink extends OutputFormatTableSink<Row> {
+
+	private final JobConf jobConf;
+	private final RowTypeInfo rowTypeInfo;
+	private final String dbName;
+	private final String tableName;
+	private final List<String> partitionColumns;
+	private final String hiveVersion;
+
+	// TODO: need OverwritableTableSink to configure this
+	private boolean overwrite = false;
+
+	public HiveTableSink(JobConf jobConf, RowTypeInfo rowTypeInfo, String dbName, String tableName,
+			List<String> partitionColumns) {
+		this.jobConf = jobConf;
+		this.rowTypeInfo = rowTypeInfo;
+		this.dbName = dbName;
+		this.tableName = tableName;
+		this.partitionColumns = partitionColumns;
+		hiveVersion = jobConf.get(HiveCatalogValidator.CATALOG_HIVE_VERSION, HiveShimLoader.getHiveVersion());
+	}
+
+	@Override
+	public OutputFormat<Row> getOutputFormat() {
+		boolean isPartitioned = partitionColumns != null && !partitionColumns.isEmpty();
+		// TODO: need PartitionableTableSink to decide
+		boolean isDynamicPartition = isPartitioned;
+		HiveTablePartition hiveTablePartition;
+		HiveTableOutputFormat outputFormat;
+		try (HiveMetastoreClientWrapper client = HiveMetastoreClientFactory.create(new HiveConf(jobConf, HiveConf.class), hiveVersion)) {
+			Table table = client.getTable(dbName, tableName);
+			StorageDescriptor sd = table.getSd();
+			// here we use the sdLocation to store the output path of the job, which is always a staging dir
+			String sdLocation = sd.getLocation();
+			if (isPartitioned) {
+				// TODO: validate partition spec
+				// TODO: strip quotes in partition values
+				LinkedHashMap<String, String> strippedPartSpec = new LinkedHashMap<>();
+				if (isDynamicPartition) {
+					List<String> path = new ArrayList<>(2);
+					path.add(sd.getLocation());
+					if (!strippedPartSpec.isEmpty()) {
+						path.add(Warehouse.makePartName(strippedPartSpec, false));
+					}
+					sdLocation = Joiner.on(Path.SEPARATOR).join(path);
+				} else {
+					List<Partition> partitions = client.listPartitions(dbName, tableName,
+							new ArrayList<>(strippedPartSpec.values()), (short) 1);
+					sdLocation = !partitions.isEmpty() ? partitions.get(0).getSd().getLocation() :
+							sd.getLocation() + Path.SEPARATOR + Warehouse.makePartName(strippedPartSpec, true);
+				}
+
+				sd.setLocation(toStagingDir(sdLocation, jobConf));
+				hiveTablePartition = new HiveTablePartition(sd, new LinkedHashMap<>(strippedPartSpec));
+			} else {
+				sd.setLocation(toStagingDir(sdLocation, jobConf));
+				hiveTablePartition = new HiveTablePartition(sd, null);
+			}
+			outputFormat = new HiveTableOutputFormat(jobConf, dbName, tableName,
+					partitionColumns, rowTypeInfo, hiveTablePartition, MetaStoreUtils.getTableMetadata(table), overwrite);
+		} catch (TException e) {
+			throw new CatalogException("Failed to query Hive metaStore", e);
+		} catch (IOException e) {
+			throw new FlinkRuntimeException("Failed to create staging dir", e);
+		}
+		return outputFormat;
```

Comment:
Nit: we can move the return statment as the finally statement in above try block.